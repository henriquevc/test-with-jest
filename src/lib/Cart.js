import find from 'lodash/find'
import remove from 'lodash/remove'

export default class Cart {

    items = []

    add (item) {
        const itemTofind = { product: item.product }

        if (find(this.items, itemTofind)) {
            remove(this.items, itemTofind)
        }

        this.items.push(item)
    }

    remove (product) {
        remove(this.items, {product})
    }

    getTotal () {
        return this.items.reduce((acc, item) => {
            return acc + item.quantity * item.product.price
        }, 0)
    }

    checkout () {
        return {
            items: this.items,
            total: this.getTotal()
        }
    }

}