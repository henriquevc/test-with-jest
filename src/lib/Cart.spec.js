import Cart from './Cart'

describe(`Cart`, () => {
    let cart
    let product = {
        title: 'chiclete bubbaloo',
        price: 35388, // 353.88
    }
    let product2 = {
        title: 'Leite condensado',
        price: 16200, // 162.00
    }

    beforeEach(() => {
        cart = new Cart()
    })

    describe('getTotal()', () => {
        it('should return 0 when getTotal() is executed in a newly created instace', () => {
            expect(cart.getTotal()).toEqual(0)
        })

        it('should multiply quantity and price and receive the total amount', () => {
            const item = {
                quantity: 2, //35388 * 2 = 70776
                product,
            }

            cart.add(item)

            expect(cart.getTotal()).toEqual(70776)
        })

        it('should ensure no more than on product exists at a time', () => {
            cart.add({
                quantity: 2,
                product,
            })
            cart.add({
                quantity: 1,
                product,
            })

            expect(cart.getTotal()).toEqual(35388)
        })

        it('should update total when a product gets included and then removed', () => {
            cart.add({
                quantity: 2,
                product,
            })

            cart.add({
                quantity: 1,
                product: product2,
            })

            cart.remove(product)

            expect(cart.getTotal()).toEqual(16200)
        })
    })

    describe('checkout()', () => {
        it('should return an object with the total and the list of items', () => {
            cart.add({
                product,
                quantity: 5,
            })

            cart.add({
                product: product2,
                quantity: 3,
            })

            expect(cart.checkout()).toMatchSnapshot(`
                Object {
                  "items": Array [
                    Object {
                      "product": Object {
                        "price": 35388,
                        "title": "chiclete bubbaloo",
                      },
                      "quantity": 5,
                    },
                    Object {
                      "product": Object {
                        "price": 16200,
                        "title": "Leite condensado",
                      },
                      "quantity": 3,
                    },
                  ],
                  "total": 225540,
                }
            `)
        })
    })
})
