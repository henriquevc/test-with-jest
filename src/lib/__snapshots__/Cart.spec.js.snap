// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart checkout() should return an object with the total and the list of items: 
                Object {
                  "items": Array [
                    Object {
                      "product": Object {
                        "price": 35388,
                        "title": "chiclete bubbaloo",
                      },
                      "quantity": 5,
                    },
                    Object {
                      "product": Object {
                        "price": 16200,
                        "title": "Leite condensado",
                      },
                      "quantity": 3,
                    },
                  ],
                  "total": 225540,
                }
             1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 35388,
        "title": "chiclete bubbaloo",
      },
      "quantity": 5,
    },
    Object {
      "product": Object {
        "price": 16200,
        "title": "Leite condensado",
      },
      "quantity": 3,
    },
  ],
  "total": 225540,
}
`;
